### ISSUE TITLE
Vectors with endpoints/ multiple points API

### ISSUE THREAD

**Issue by sofroniewn** [napari/napari#4009]
## 🚀 Feature

Should we expand the vectors API to add

1. Support for passing start point and end point as an `Nx2xD` array. Right now we only support start point + vector. We'd need to add a kwarg such as `as_points` to the API.

2.  Support for passing multiple points as an `NxCxD` array. Where `C` represents the length of the longest multiline. If you wanted to have some lines longer than others you'd need to use `NaN` to indicate that.

For reference here is current API

https://github.com/napari/napari/blob/410459a45db30efdf84c2339db9ceff8ccc570c8/napari/layers/vectors/vectors.py#L26-L31

## Motivation

For 1. sometimes you have data in a start + end point format and converting to a start point + vector format feels annoying / wasteful / a burden

For 2. Right now if you want to display multilines / paths you have to use the shapers layer which has more interactivity but is not as performant, and struggles for datasets will millions of lines. This difference could be even greater after #4005 merges

## Additional context

What does @brisvag think? Also curious @alisterburt @kevinyamauchi @jni @andy-sweet and anyone else who has worked with vectors

**Comment by brisvag** [comment]
> 1. Support for passing start point and end point as an `Nx2xD` array. Right now we only support start point + vector. We'd need to add a kwarg such as `as_points` to the API.

YES, this is great and I don't know why I haven't proposed it earlier, since it seems pretty trivial to implement. @alisterburt  and I have had to juggle with this many times. Support 100%! The only obstacle is the api extension, but I think it's worth it. Something that we need to think about is where the `length` parameter should go in this new world.

> 2. Support for passing multiple points as an `NxCxD` array. Where `C` represents the length of the longest multiline. If you wanted to have some lines longer than others you'd need to use `NaN` to indicate that.

This seems pretty cool as well. This doesn't even require an api extension, since it's just a generalization of the current implementation.

**Comment by andy-sweet** [comment]
> For 1. sometimes you have data in a start + end point format and converting to a start point + vector format feels annoying / wasteful / a burden

My assumption is that this will be converted/coerced internally though, likely by the existing `fix_data_vectors` function? I.e. the stored type/shape of `Vectors.data` is the `Nx2xD` array that represents the start point and vector and won't change, right?

I think that's fine. The API expansion is only as a parameter for `__init__`, so I also think that's OK. Alternatively, we could define a public utility function that the user has to call, but that may be less discoverable.

> For 2. Right now if you want to display multilines / paths you have to use the shapers layer which has more interactivity but is not as performant, and struggles for datasets will millions of lines. This difference could be even greater after #4005 merges

My initial gut reaction is that `Vectors` is a bad name for this layer now, and something like `Lines` or `Paths` would be better. But maybe this is OK as a temporary workaround if there's a need for it. I'm unsure exactly what the `Graphs` layer is going to be, but I could also imagine a version of that as a better long term generalization of these types of things.

**Comment by alisterburt** [comment]
> Support for passing start point and end point as an Nx2xD array. Right now we only support start point + vector. We'd need to add a kwarg such as as_points to the API.

I like the utility here but worry that having the same data render different lines might lead to some confusion... the API isn't super intuitive and requires a user to know/understand both possibilities to use the layer properly

> Support for passing multiple points as an NxCxD array. Where C represents the length of the longest multiline. If you wanted to have some lines longer than others you'd need to use NaN to indicate that.

This is really great - multilines would be a really nice feature to add

> My initial gut reaction is that Vectors is a bad name for this layer now

This is my feeling too. Supporting line/multiline in the vectors layer feels a bit strange - would having the line/multiline visual backing the shapes layer in certain situations be a possibility? this functionality feels like a more natural fit there... maybe we could have a kwarg and disable interactive controls if rendering a large number of lines/multilines?

Requiring different representations for fast visualisation vs interactive editing has come up for the tracks layer (without a clear solution) too in #3896
